### ISSUE TITLE
Make simplistic builds more viable

### ISSUE THREAD

**Issue by busstoptaktik** [OSGeo/PROJ#622]
**TL;DR:** I believe we should go slightly out of our way to support config-less build methods. Hence, I suggest we introduce some small modifications that simplify hand held builds significantly. @aaronpuchert, @kbevers, @hobu, @rouault and @QuLogic: I mention you here, since you may have opinions and/or advice regarding this.

(This grew from a discussion with @aaronpuchert over at PR #618)

PROJ.4 is basically a very simple library, that you should be able to build on even very modest platforms. I think we should keep it that way.

I recently did an experiment (for reference, let’s call it “the simplistic build”): I took everything in the proj.4/src directory and splitted up in 4:

`proj.4/src/apps` with everything having a “main” function (proj.c, cs2cs.c etc.)
`proj.4/src/lib` with all other c files
`proj.4/src/include` with all header files
`proj.4/src/etc` with everything else

That done, from `proj.4/src`, I said:
```
gcc -I./include -c lib/*.c
```
Which revealed that `proj_config.h` was used by only 2 (3) source files: `jniproj.c`, `pj_strtod.c`, and, on non-windows platforms, `pj_mutex.c`. For now, I could live without `jniproj.c`, and just deleted it. Then I added functions to `proj_strtod.c`, forwarding calls to `pj_strtod` and `pj_atof` to my own `proj_strtod` and `proj_atof`, and deleted `pj_strtod.c`

Now, I was left with a proj_config.h independent library source tree, so I said:
```
gcc -I./include -c lib/*.c
ar r libproj4.a lib/*.o
ar s libproj4.a
gcc -I./include -L. -o gie apps/gie.c -lproj4 -lm
gie ../test/gie/builtins.gie
```
Which answered with:
```
-------------------------------------------------------------------------------
Reading file '..\test\gie\builtins.gie'
-------------------------------------------------------------------------------
total: 1306 tests succeeded,   0 tests failed.
-------------------------------------------------------------------------------
```
The remaining apps (proj, cs2cs, cct) also built flawlessly, and actually `proj_config.h` is hardly used for anything:
```
proj.4/ $ grep -r proj_config *

ChangeLog:	of proj_config header, and shut off Framework building by
cmake/Makefile.am:			proj_config.cmake.in \
cmake/Makefile.am:			proj_config.cmake.in \
cmake/Proj4Config.cmake:configure_file(cmake/proj_config.cmake.in src/proj_config.h)
configure.ac:AM_CONFIG_HEADER(src/proj_config.h)
src/jniproj.c:#include "proj_config.h"
src/lib_proj.cmake:        ${CMAKE_CURRENT_BINARY_DIR}/proj_config.h
src/lib_proj.cmake:## targets: libproj and proj_config.h
src/pj_mutex.c:#include "proj_config.h"
src/pj_strtod.c:/* Windows nmake build doesn't have a proj_config.h, but HAVE_LOCALECONV */
src/pj_strtod.c:#include "proj_config.h"
src/proj_config.h.wince:# error This version of proj_config.h header is dedicated for Windows CE platform!
wince/msvc80/projce_dll/projce_dll.vcproj:        <File RelativePath="..\..\..\src\proj_config.h">
wince/msvc80/projce_lib/projce_lib.vcproj:        <File RelativePath="..\..\..\src\proj_config.h">
```
jniproj.c uses proj_config.h to check whether it should do anything at all:

    #include "proj_config.h"
    #ifdef JNI_ENABLED
    ...
    #endif

and `pj_mutex.c` uses it (on non-Windows platforms) for a single capability check:
```
#ifdef HAVE_PTHREAD_MUTEX_RECURSIVE
        pthread_mutexattr_settype(&mutex_attr, PTHREAD_MUTEX_RECURSIVE);
#else
        pthread_mutexattr_settype(&mutex_attr, PTHREAD_MUTEX_RECURSIVE_NP);
#endif
```
So, if we can live with this fallback to non-recursive mutexes on POSIX platforms, it would be a simple thing to support a plain handheld (or makefile) style build, by adding a few conditionals in `pj_mutex.c`, `pj_strtod.c` and `jniproj.c`

Given how much effort we put into supporting the cmake, nmake and autotools based build systems, I believe it would be worth the little extra effort to support more simplistic builds.

Now, since I’m not a build system wizard, I need some advice about how to do it best. I *think* autotools defines the `HAVE_CONFIG_H` symbol, and I *think* we could force cmake to do that too. If this is true, the conditionals could be based on that.

Otherwise, we could force simplistic builds to actively tell what they want by defining e.g. `NO_CONFIG_H`. I would, however, prefer to have the (slight) effort done by the automatic build systems, and not put the burden on the hand held builder.

Also, I do not suggest immediately to split up the source code the way I did under the simplistic build. I think it would be a good idea to separate the library code from the include files and the command line apps, but actually doing it would also imply a large number of corrections in the build system support files, that I wouldn’t volunteer to do for now. Rather, I would probably just make a few comments about simplistic builds in INSTALL or README.

Additionally, this is *not* a suggestion for ditching the existing build system, it is just a call for ensuring that PROJ.4 remains buildable on systems having nothing but a C89 compiler, and, perhaps, `make`. I believe a simplistic build system would also lower the barrier of entry, and hence make PROJ.4 more inviting for new contributors.

**Comment by QuLogic** [comment]
There was a long discussion of this on the mailing list earlier this year, and from my understanding of [the summary](http://lists.maptools.org/pipermail/proj/2017-June/007725.html) by @kbevers, `nmake` support would be dropped to focus on CMake and autotools. (Personally, I have a preference for Meson, but I understand that's pretty new overall.) I'm not sure adding another build system would help this new focus (even if it's "simple").

**Comment by aaronpuchert** [comment]
I think the idea is to not rely on any build system too much and to keep the number of required configuration `#define`s as small as possible. This would make future migrations a lot easier.

> I took everything in the `proj.4/src` directory and splitted up in 4:
`proj.4/src/apps` with everything having a “main” function (`proj.c`, `cs2cs.c` etc.)
`proj.4/src/lib` with all other c files
`proj.4/src/include` with all header files
`proj.4/src/etc` with everything else

I would welcome such a change, with some modifications maybe:
- `proj.4/src/tools` with everything having a “main” function, i.e. code that doesn't go into the library,
- `proj.4/src/lib` containing the library implementation,
- `proj.4/include` for public headers, i.e. the API for users of the library,
- `proj.4/src/include` for internal headers,
- `proj.4/src/tests` for tests, i.e. something like `pj_generic_selftest.c` and `pj_run_selftests.c`.

Splitting up the `include` directories would be in the spirit of offering a new API in the future.

Maybe the implementation could even be split up further into
- the core functionality, i.e. all projections,
- additional functionality like `mk_cheby.c` or `geodesic.c`.

**Comment by busstoptaktik** [comment]
@QuLogic said:

> Personally, I have a preference for Meson, but I understand that's pretty new overall.

The combination of Meson and Ninja is a beauty, and I would enjoy seeing it succeed in conquering the world through its undeniable virtue of providing a *tool for programmers* supporting a syntax that can actually be *read by programmers* (this also implies that I would be excited to see a proj.4/src/meson.build file, despite all attempts to keep the number of build/source code configuration systems low). 

For now, for building PROJ.4, I use the CMAKE Ninja backend, but the CMAKE syntax is not inviting, and I really appreciate that wiser people than myself have done the heavy lifting by authoring the PROJ.4 CMAKE files initially.

> I'm not sure adding another build system would help this new focus (even if it's "simple").

Please reread my comment. I'm ***not*** proposing to add a new build system. I'm proposing (as @aaronpuchert also correctly interprets the text) to strength reduce an unneccessary ***hard*** dependency on a ***source code configuration system***, to a ***soft*** dependency.

Since the role of the source code configuration system during build is so limited in the case of PROJ.4, my suggestion was (and is) to take sufficient action to ensure that *we do not step on the toes* of people needing to build PROJ.4 under less common circumstances (e.g. on embedded hardware platforms, or when needing to embed/"vendoring in" PROJ.4 as part of a larger software system).

So I'm *not* talking about adding build systems - I'm talking about not letting source code configuration systems get in the way of people needing to handhold their build process.

**Comment by QuLogic** [comment]
You're right that the effects of a configuration system are quite minimal in Proj.4. Instead of `NO_CONFIG_H`, maybe `HAVE_CONFIG_H` instead, which is only defined when using one of these systems. The options from the file would default to something sane (which they pretty much already do) unless the builder defined the relevant options themselves.

And since config is so straightforward, here's a side-project of a [meson port](https://github.com/QuLogic/proj.4/tree/meson). It doesn't quite support all the nad stuff due to all the wildcards that I didn't want to spend too much time on. I also think Java headers could be autogenerated (in autotools/cmake too), but didn't really look into it. Also, this was based on the autotools build which I think seems to be missing quite a few non-essential things that the cmake build does.

**Comment by mloskot** [comment]
@busstoptaktik
> reduce an unneccessary hard dependency on a source code configuration system, to a soft dependency.

AFAIU, this is a call for maintaining real portability at a source code level, independent from any build system configurations. IMHO, this is desirable. "Don't let yourself be fooled into popularity of build systems."

**Comment by busstoptaktik** [comment]
In this context it will also be relevant to enable support for [unity builds](http://mesonbuild.com/Unity-builds.html).

It mostly requires prefixing of repeated static function names (s_forward, s_inverse etc.) with projection mnemonics.

This will make it simple to handle libproj as part of a larger project simply by providing one huge concatenated .c file, which (due to being one compilation unit) may even get better overall optimization out of the box (sqlite works this way).

**Comment by aaronpuchert** [comment]
Unity builds are probably going to stumble over all the macro magic used throughout the code base. I don't think it's worth the trouble. The only files meant to be `#include`d are header files, and there are better ways to get optimizations across compile units.

The code is also not terribly heavy on header files, so I don't see why unity builds should be faster. (Their advantage is that the compiler frontend needs to parse header files only once.)

**Comment by busstoptaktik** [comment]
@aaronpuchert said:

> Unity builds are probably going to stumble over all the macro magic.

Actually, there is not much macro magic left compared to what was there when @kbevers and I started working on the code base 2 years ago.

Also, see my reply to the comment by @kbevers over in #841, where I outline a plan for eliminating the last parts (the `OPERATION` macro and its specializations, and the [X macro magic](https://en.wikipedia.org/wiki/X_Macro) in `pj_list.c`/`pj_list.h`.

> The code is also not terribly heavy on header files, so I don't see why unity builds should be faster. (Their advantage is that the compiler frontend needs to parse header files only once.)

I'm not talking about faster compilation, I'm talking about potentially faster execution, since the compiler, at high optimization levels, is free to do agressive inlining, because the entire library becomes one single object file.

Also, please note that I do not personally see this as a major advantage, but apparently [`sqlite`](http://sqlite.org) has had success with this way of distributing things, since it makes it easy for projects consuming the library to stick to a frozen version by including just one source and one header file in their code base.

I do not necessarily think this is neither good nor perfect, but essentially I would leave the definition of perfection to the user and would try to avoid stepping on anyones toes wrt. whichever way they would want to do things: I don't think we should distribute libproj as a single file, but i would not want to stand in the way for anyone wanting to do

    cat src/proj_internal.h src/*.c > libproj.c

and then including `libproj.c` and `proj.h` in their project code base as a frozen libproj version.

As you mention, there are other ways to do this - and this is just one more. It's not the right way for neither you nor me, but it may be for Hansel or Gretel.

**Comment by aaronpuchert** [comment]
There are certainly less problems with macros now than in the past, but I still remember [how I struggled](https://github.com/OSGeo/proj.4/pull/674#issuecomment-350266668) to add a single header (`proj.h`) to some source files. This should be easy, and in most code bases it is. These problems are still there. (I'm talking about `PJ_LIB__` and and the very liberal usage of `PROJ[_INTERNAL]_H` beyond their role as header guards.)

Certainly the new API is going to improve a lot in that regard, but the old [isn't going away soon](https://github.com/OSGeo/proj.4/issues/686#issuecomment-347311321).

SQLite does indeed recommend the usage of their amalgamation, but on Linux I can just install `sqlite3-dev[el]` and then I have `sqlite3.h` to include and `libsqlite3.so` to link. I'm not sure why I would want to compile my own SQLite for every project I use it in. This makes no sense to me.

Now about optimization. Compilation consists of four steps:

1. frontend (parsing the source code),
2. optimization,
3. backend (machine code generation), and
4. linking,

usually in that order. For optimization over module boundaries it would be necessary to link *before* optimization.

Concatenating source files means doing the linking first and the other three steps after that. But C source files are not meant to be concatenated. (Header files are a different story, it should be possible to textually include them. This is why they have a different file ending.) With concatenation, things like [internal linkage](http://en.cppreference.com/w/c/language/storage_duration) (i.e. `static`) don't work anymore, meaning all symbols are more or less global. This is a serious limitation, it means I can't just write valid C code anymore and be done with it.

Wouldn't it be easier to do the linking between frontend and optimization? Then the frontend sees the original source code, and there are no restrictions on what I can do. That is what LTO (link-time optimization) does. With Clang/LLVM one can do it simply by plugging the tools together: compile with `clang -c -emit-llvm`, then link using `llvm-link`, optimize with `opt` and generate the machine code with `llc`. (Of course [in practice](http://blog.llvm.org/2016/06/thinlto-scalable-and-incremental-lto.html) one would use `-flto`, which is supported by both [GCC](https://gcc.gnu.org/wiki/LinkTimeOptimization) and [Clang](https://llvm.org/docs/LinkTimeOptimization.html).) No hacks required. No problems with static linkage. This is how it's meant to be.

Unity builds are a dirty hack, and in my opinion any development effort on supporting them is wasted time.

**Comment by aaronpuchert** [comment]
By the way, supporting builds like `gcc -o proj -include src/ src/*.c` is completely fine for me. We don't have any dependencies, apart from the standard library, so autotools and CMake are probably a bit overkill.

**Comment by busstoptaktik** [comment]
As @aaronpuchert summed it up:

> supporting builds like gcc -o proj -include src/ src/*.c is completely fine for me. We don't have any dependencies, apart from the standard library

This is, however, no longer the case, so closing this as "non-viable"
